### PR TITLE
Hardware components extension for lifecycle support

### PR DESCRIPTION
**Pull request by destogl**
This is proposal for #107 (check the last tree comments there also).

Compoentns are now using the same structre as controller for the lifecycle.﻿

### CODE REVIEW

File: hardware_interface/include/hardware_interface/actuator.hpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (Karsten1987) [inline hardware_interface/include/hardware_interface/actuator.hpp:1]: The copyright should always state the original date of authoring that file.
alternatively you could specify a range of the file, that is `2020 - 2021`. 

Same comment for the files below.

File: hardware_interface/include/hardware_interface/actuator_interface.hpp [hunk 2]
```diff
@@ -22,7 +22,10 @@
 #include "hardware_interface/handle.hpp"
 #include "hardware_interface/hardware_info.hpp"
 #include "hardware_interface/types/hardware_interface_return_values.hpp"
-#include "hardware_interface/types/hardware_interface_status_values.hpp"
+#include "rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp"
```
Comment (Karsten1987) [inline hardware_interface/include/hardware_interface/actuator_interface.hpp:25]: why not inheriting from the `lifecycle_node_interface`? That way we have all the function signatures for free (and their default implementation as well if wanted).

File: hardware_interface/CMakeLists.txt [hunk 3]
```diff
@@ -36,6 +37,7 @@ ament_target_dependencies(
   hardware_interface
   control_msgs
   pluginlib
+  rclcpp_lifecycle
```
Comment (Karsten1987) [inline hardware_interface/CMakeLists.txt:40]: I am not sure if we really want that dependency on the hardware layer.

**Review by Karsten1987 — COMMENTED** [review]
I am a bit hesitant to merge the `rclcpp_lifecycle` dependency on the hardware interfaces. So far it's been a relatively lean setup with mostly leaf dependencies such as `rcutils` etc. With the lifecycle nodes we get the full stack of ROS2 attached to the hardware interfaces.

Additionally, I am not sure exactly what we win by this addition. Couldn't we just emulate the lifecycle API ourselves? Looking at the current change, there's not too much leverage taken. I guess if we inherited from the lifecycle interfaces, we'd get the declarations and their implementations for free which makes the diff smaller.

File: hardware_interface/CMakeLists.txt (hunk 3 above)
Comment (destogl) [reply hardware_interface/CMakeLists.txt:35]: Is this possible, since we are inheriting from `LifecycleNodeInterface`?

**Comment by destogl** [comment]
@Karsten1987, @bmagyar this is ready for (harsh) review :D

File: hardware_interface/include/hardware_interface/actuator.hpp (hunk 1 above)
Comment (destogl) [reply hardware_interface/include/hardware_interface/actuator.hpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/include/hardware_interface/actuator_interface.hpp [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/actuator_interface.hpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/include/hardware_interface/sensor_interface.hpp [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/sensor_interface.hpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/include/hardware_interface/system.hpp [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/system.hpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/src/sensor.cpp [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/src/sensor.cpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/src/system.cpp [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/src/system.cpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/include/hardware_interface/sensor.hpp [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/sensor.hpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/include/hardware_interface/system_interface.hpp [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 ros2_control Development Team
+// Copyright 2021 ros2_control Development Team
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/system_interface.hpp:1]:
```suggestion
// Copyright 2020 - 2021 ros2_control Development Team
```

File: hardware_interface/include/hardware_interface/types/lifecycle_state_names.hpp [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+/// \author: Denis Stogl
+
+#ifndef HARDWARE_INTERFACE__TYPES__LIFECYCLE_STATE_NAMES_HPP_
```
[the remark this answers is not in the document]
Comment (destogl) [reply hardware_interface/include/hardware_interface/types/lifecycle_state_names.hpp:17]: referenced in #536

File: hardware_interface/include/hardware_interface/types/lifecycle_state_names.hpp [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+#define HARDWARE_INTERFACE__TYPES__LIFECYCLE_STATE_NAMES_HPP_
+
+#include <string>
+
+namespace hardware_interface
+{
+namespace lifecycle_state_names
```
[the remark this answers is not in the document]
Comment (destogl) [reply hardware_interface/include/hardware_interface/types/lifecycle_state_names.hpp:24]: will be done in #537

**Review by bmagyar — CHANGES_REQUESTED** [review]

**Review by bmagyar — APPROVED** [review]

**Pull request merged**